fdm-core/src/authorization.test.ts

```diff
@@ -0,0 +1,366 @@
+            await grantRole(fdm, "farm", "owner", farm_id, principal_id)
+            await checkPermission(
+                fdm,
+                "farm",
+                "read",
+                farm_id,
+                principal_id,
+                "test",
+            )
+
+            const auditLogs = await fdm
+                .select()
+                .from(authZSchema.audit)
+                .where(eq(authZSchema.audit.principal_id, principal_id))
+                .orderBy(desc(authZSchema.audit.audit_timestamp))
+            expect(auditLogs.length).toBeGreaterThanOrEqual(1)
+            expect(auditLogs[0].target_resource).toBe("farm")
+            expect(auditLogs[0].target_resource_id).toBe(farm_id)
+            expect(auditLogs[0].action).toBe("read")
+            expect(auditLogs[0].allowed).toBe(true)
+        })
+
+        it("should store the audit log when a permission check is performed and not allowed", async () => {
+            const principal_id_new = createId()
+
+            await expect(
+                checkPermission(
+                    fdm,
+                    "farm",
+                    "read",
+                    farm_id,
+                    principal_id_new,
+                    "test",
+                ),
+            ).rejects.toThrowError(
+                "Principal does not have permission to perform this action",
+            )
+
+            const auditLogs = await fdm
+                .select()
+                .from(authZSchema.audit)
+                .where(eq(authZSchema.audit.principal_id, principal_id_new))
+                .orderBy(desc(authZSchema.audit.audit_timestamp))
+            expect(auditLogs.length).toBeGreaterThanOrEqual(1)
+            expect(auditLogs[0].target_resource).toBe("farm")
+            expect(auditLogs[0].target_resource_id).toBe(farm_id)
+            expect(auditLogs[0].action).toBe("read")
+            expect(auditLogs[0].allowed).toBe(false)
+        })
+    })
+
+    describe("grantRole", () => {
+        it("should grant a role to a principal for a resource", async () => {
+            await grantRole(fdm, "farm", "owner", farm_id, principal_id)
+
+            const roles = await fdm
+                .select()
+                .from(authZSchema.role)
+                .where(
+                    and(
+                        eq(authZSchema.role.resource, "farm"),
+                        eq(authZSchema.role.resource_id, farm_id),
+                        eq(authZSchema.role.principal_id, principal_id),
+                        eq(authZSchema.role.role, "owner"),
+                        isNull(authZSchema.role.deleted),
+                    ),
+                )
+            expect(roles.length).toBe(1)
+        })
+
+        it("should handle non-unique role grants", async () => {
+            await grantRole(fdm, "farm", "owner", farm_id, principal_id)
+            await grantRole(fdm, "farm", "owner", farm_id, principal_id)
+
```

Comment:
why this duplication, you grant it twice?

Reply:
This test confirms that if an user had a role and it is assigned again that role it still available

Reply:
ok. that makes sense.